keras_nlp/layers/sine_positional_embedding.py

```diff
@@ -0,0 +1,92 @@
+# Copyright 2022 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Sinusoidal Position embedding implementation based on
+`keras.layers.Layer`."""
+
+import tensorflow as tf
+from tensorflow import keras
+
+
+class SinePositionEncoding(keras.layers.Layer):
+    """Creates a layer which ouputs a sinusoidal positional embedding for
+    inputs sequences
+
+    This layer calculates the position encoding as a mix of sine and cosine
+    functions with geometrically increasing wavelengths. Defined and formulized
+    in "Attention is All You Need", section 3.5.
+    (https://arxiv.org/abs/1706.03762).
+
+    Args:
+      min_frequency: The minimum frequency for each position embedding,
+                     initalized to 1.0e-4.
+
+    Example:
+    ```python
+    positional_embedding = keras_nlp.layers.SinePositionEncoding()
+    inputs = keras.Input((100, 32), dtype=tf.float32)
+    outputs = positional_embedding(inputs)
+    ```
+
+    References:
+        [Vaswani et al., 20XX](https://arxiv.org/abs/1706.03762)
+
+    """
+
+    def __init__(self, min_frequency: float = 1.0e-4, **kwargs):
+
+        super().__init__(**kwargs)
+        self.min_frequency = min_frequency
+
+    def get_config(self):
+        config = super().get_config()
+        config.update(
+            {
+                "min_frequency": self.min_frequency,
+            }
+        )
+        return config
+
+    def call(self, inputs):
+        """Forward pass of the SinePositionEncoding.
+
+        Args:
+            inputs: a Tensor. The length along the second last dimension is the
+            length of the sequence.
+
+        Returns:
+            A Tensor of the same shape as the `inputs` containing the sinusoidal
+            positional embedding weights for the input sequence.
+        """
+        input_shape = tf.shape(inputs)
+        # length of sequence is the second last dimension of the inputs
+        seq_length = input_shape[-2]
+        hidden_size = input_shape[-1]
+        min_freq = self.min_frequency
+        position = tf.range(seq_length, dtype=tf.float32)
```

Comment:
dtype should probably default to `self.compute_dtype` to support mixed precision